multipass-addressing/src/main/scala/com/precisely/bigdata/sample/spark/MultipassAddressing.scala

```diff
@@ -0,0 +1,198 @@
+
+  import com.precisely.addressing.v1.model.Response
+  import com.precisely.addressing.v1.{Addressing, Preferences}
+
+  // set preference to return all fields and Match Mode to Relaxed
+  private val relaxedGeocoderPreferences: Preferences = new PreferencesBuilder()
+    .withReturnAllInfo(true)
+    .withMatchMode("RELAXED")
+    .build()
+
+  override def execute(input: RequestInput, preferences: Option[Preferences], addressing: Addressing): Response = {
+
+    val verifyResponse = addressing.verify(input.requestAddress(), preferences.orNull)
+
+    val inputRequest: RequestAddress = input.requestAddress()
+
+    var multiLineRequest: RequestAddress = inputRequest
+
+    if (CollectionUtils.isNotEmpty(verifyResponse.getResults)) {
+      val verifyAddress = verifyResponse.getResults.get(0).getAddress
+
+      //  create multiLine request from verify Response
+      multiLineRequest = new RequestAddress()
+      multiLineRequest.setAddressLines(getValidAddressLine(verifyResponse.getResults.get(0).getAddressLines))
+      multiLineRequest.setStreet(verifyAddress.getStreet)
+      multiLineRequest.setAddressNumber(verifyAddress.getAddressNumber)
+      multiLineRequest.setCity(verifyAddress.getCity.getLongName)
+      multiLineRequest.setPostalCode(verifyAddress.getPostalCode)
+      multiLineRequest.setCountry(verifyAddress.getCountry.getIsoAlpha3Code)
+
+    }
+
+    val multiLineResponse = addressing.geocode(multiLineRequest, relaxedGeocoderPreferences)
+    if (CollectionUtils.isNotEmpty(multiLineResponse.getResults) && isAddressLevelMatch(multiLineResponse.getResults.get(0))) {
+      return multiLineResponse
+    }
+
+    val singleLine: String = if (CollectionUtils.isNotEmpty(verifyResponse.getResults)) {
+      for (addressLine <- 0 to getValidAddressLine(verifyResponse.getResults.get(0).getAddressLines).size() - 1) {
+        verifyResponse.getResults.get(0).getAddressLines.get(addressLine) + " "
+      }
+      verifyResponse.getResults.get(0).getAddress.getFormattedAddress + " " + verifyResponse.getResults.get(0).getAddress.getCountry.getIsoAlpha3Code
+    } else {
+      inputRequest.getStreet + " " +
+        inputRequest.getCity + " " +
+        inputRequest.getAdmin1 + " " +
+        inputRequest.getPostalCode
+    }
+
+    val singleLineRequest: RequestAddress = new RequestAddress()
+    singleLineRequest.setAddressLines(Collections.singletonList(singleLine))
+
+    val singleLineResponse = addressing.geocode(singleLineRequest, relaxedGeocoderPreferences)
+    if (CollectionUtils.isNotEmpty(singleLineResponse.getResults) && isAddressLevelMatch(singleLineResponse.getResults.get(0))) {
+      return singleLineResponse
+    }
+
+    if (CollectionUtils.isNotEmpty(multiLineResponse.getResults)) {
+      if (CollectionUtils.isNotEmpty(singleLineResponse.getResults)
+        && singleLineResponse.getResults.get(0).getScore >= multiLineResponse.getResults.get(0).getScore) {
+        return singleLineResponse
+      }
```

Comment:
Minor change here: the operator should be '>' rather than '>='
The idea that I presented in that previous content was that we return the single line response only if the score was strictly better than the multi line score.  Because we don't know (in the general case) if the user is pulling other fields from the response, we want to be careful to return the multi-line geocode by default in the case of a tie.  It would probably be close or identical in most cases, but i think its cleaner to think of it as us preferring the multi-line and only taking single line when clearly better.
Also, add a comment, something like: "Return the single line response only if it is better than the multi line response"

Reply:
Ok